source/agora/consensus/state/ValidatorSet.d

```diff
@@ -579,6 +579,48 @@ public class ValidatorSet
         return PreImageInfo.init;
     }
 
+    /***************************************************************************
+
+        Get validators' pre-image information
+
+        Params:
+            start_height = the starting enrolled height to begin retrieval from
+            end_height = the end enrolled height to finish retrieval to
+
+        Returns:
+            preimages' information of the validators
+
+    ***************************************************************************/
+
+    public PreImageInfo[] getPreimages (ulong start_height,
+        ulong end_height) @trusted nothrow
+    {
+        PreImageInfo[] preimages;
+
+        try
+        {
+            auto results = this.db.execute("SELECT key, preimage, distance " ~
+                "FROM validator_set WHERE enrolled_height >= ? AND " ~
+                "enrolled_height <= ?",
+                start_height, end_height);
+
+            foreach (row; results)
+            {
+                Hash enroll_key = Hash(row.peek!(char[])(0));
+                Hash preimage = Hash(row.peek!(char[])(1));
+                ushort distance = row.peek!ushort(2);
+                preimages ~= PreImageInfo(enroll_key, preimage, distance);
+            }
+        }
+        catch (Exception ex)
+        {
+            log.error("Exception occured on getPreimages: {}, heights " ~
+                "[{}..{}]", ex.msg, start_height, end_height);
+        }
+
+        return preimages;
```

Comment:
You could consider using fold to build the array directly from the results:-
```
public PreImageInfo[] getPreimages (ulong start_height,
        ulong end_height) @trusted nothrow
    {
        PreImageInfo[] empty;
        try
        {
            auto results = this.db.execute("SELECT key, preimage, distance " ~
                "FROM validator_set WHERE enrolled_height >= ? AND " ~
                "enrolled_height <= ?",
                start_height, end_height);

            return results.fold!((preimages, row) =>
                preimages ~= PreImageInfo(
                    Hash(row.peek!(char[])(0)),
                    Hash(row.peek!(char[])(1)),
                    row.peek!ushort(2)))(empty);
        }
        catch (Exception ex)
        {
            log.error("Exception occured on getPreimages: {}, heights " ~
                "[{}..{}]", ex.msg, start_height, end_height);
        }

        return empty;
    }
```

Reply:
Good suggestion, thanks.

Reply:
The code you suggested outputs some error like this.

> Error: no property fold for type d2sqlite3.results.ResultRange

I have to learn more about the`fold`. It's not bad to leave the code as it is but I promise I will use the `fold` in a similar context like this one. Thanks for introducing the new concept.

Reply:
I think you probably missed the import std.algorithm

Reply:
Sorry I forgot to include in the sugestion.

Reply:
This does work:-
```
public PreImageInfo[] getPreimages (ulong start_height,
        ulong end_height) @trusted nothrow
    {
        import std.algorithm;
        PreImageInfo[] empty;
        try
        {
            auto results = this.db.execute("SELECT key, preimage, distance " ~
                "FROM validator_set WHERE enrolled_height >= ? AND " ~
                "enrolled_height <= ?",
                start_height, end_height);

            return results.fold!((preimages, row) =>
                preimages ~= PreImageInfo(
                    Hash(row.peek!(char[])(0)),
                    Hash(row.peek!(char[])(1)),
                    row.peek!ushort(2)))(empty);
        }
        catch (Exception ex)
        {
            log.error("Exception occured on getPreimages: {}, heights " ~
                "[{}..{}]", ex.msg, start_height, end_height);
        }

        return empty;
    }
```

Reply:
No need to change now but just for reference.

Reply:
Right. I have already succeeded with your information yesterday. :) But, I decided to use the original code because using `foreach` is easier to understand, I think. I will apply the `fold` in more adequate situation or context. Thanks.